log/include/gz/utils/log/SplitSink.hh

```diff
@@ -0,0 +1,213 @@
+
+#ifndef GZ_UTILS_LOG_SPLITSINK_HH_
+#define GZ_UTILS_LOG_SPLITSINK_HH_
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <spdlog/details/log_msg.h>
+#include <spdlog/details/null_mutex.h>
+#include <spdlog/pattern_formatter.h>
+#include <spdlog/sinks/base_sink.h>
+#include <spdlog/sinks/ringbuffer_sink.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+
+namespace gz::utils::log
+{
+/// \brief Logging sink for spdlog that logs in Gazebo-conventions.
+///
+/// This will route messages with severity (warn, err, critical) to stderr,
+/// and all other levels (info, debug, trace) to stdout.
+template<typename Mutex>
+class SplitConsoleSink : public spdlog::sinks::base_sink<Mutex>
+{
+  /// \brief Class constructor.
+  public: SplitConsoleSink() = default;
+
+  /// \brief No copy constructor.
+  public: SplitConsoleSink(const SplitConsoleSink &) = delete;
+
+  /// \brief No asignment operator.
+  public: SplitConsoleSink &operator=(const SplitConsoleSink &) = delete;
+
+  /// \brief Try to log a message.
+  /// \param[in] _msg The message to log.
+  protected: void sink_it_(const spdlog::details::log_msg &_msg) override
+  {
+    if (!this->should_log(_msg.level))
+      return;
+
+    if (_msg.level == spdlog::level::warn      ||
+        _msg.level == spdlog::level::err       ||
+        _msg.level == spdlog::level::critical)
+    {
+      this->stderr.log(_msg);
+    }
+    else
+      this->stdout.log(_msg);
+  }
+
+  /// \brief Flush messages.
+  protected: void flush_() override
+  {
+    this->stdout.flush();
+    this->stderr.flush();
+  }
+
+  /// \brief Set the logging pattern.
+  /// \param[in] _pattern The logging pattern.
+  protected: void set_pattern_(const std::string &_pattern) override
+  {
+    this->set_formatter(
+      spdlog::details::make_unique<spdlog::pattern_formatter>(_pattern));
+  }
+
+  /// \brief Set the new formatter.
+  /// \param[in] _sinkFormatter The formatter.
+  protected: void set_formatter_(
+    std::unique_ptr<spdlog::formatter> _sinkFormatter) override
+  {
+    spdlog::sinks::base_sink<Mutex>::formatter_ = std::move(_sinkFormatter);
+    this->stdout.set_formatter(
+      spdlog::sinks::base_sink<Mutex>::formatter_->clone());
+    this->stderr.set_formatter(
+      spdlog::sinks::base_sink<Mutex>::formatter_->clone());
+  }
+
+  /// \brief Standard output.
+  private: spdlog::sinks::stdout_color_sink_st stdout;
+
+  /// \brief Standard error.
+  private: spdlog::sinks::stderr_color_sink_st stderr;
+};
+
+using SplitConsoleSinkMt = SplitConsoleSink<std::mutex>;
+using SplitConsoleSinkSt = SplitConsoleSink<spdlog::details::null_mutex>;
+
+/// \brief Logging sink for spdlog that logs in Gazebo-conventions
+///
+/// This will route messages with severity (warn, err, critical) to stderr,
+/// and all other levels (info, debug, trace) to stdout
+template<typename Mutex, size_t numItems>
+class SplitRingBufferSink: public spdlog::sinks::base_sink<Mutex>
```

Comment:
Can `numItems` be a constructor argument instead of a template? I only see it used to initialize the data members `stdout` and `stderr`. If that's the only reason, that could be done from the constructor.

Reply:
@mjcarroll , do you have an opinion here?

Reply:
This was also following an upstream version, but looking at the most recent version, it seems they have dropped this as a template and it's only a constructor argument: https://github.com/gabime/spdlog/blob/v1.x/include/spdlog/sinks/ringbuffer_sink.h#L20-L24

I'm good with dropping the template here as well.

Reply:
Not needed anymore.